GoogleForCreators/web-stories-wp
# Telemetry: Add foundation for anonymous usage tracking

## Summary

This PR introduces **opt-in**, Google Analytics-based usage tracking for all admin screens we interact with.

**Screens Affected**

* Stories Dashboard
* Stories Editor
* Block Editor (for the embed block)

**How to opt-in**

**By default, no tracking occurs.** To opt-in, modify `\Google\Web_Stories\Tracking::is_active()` in the plugin, filter `get_user_metadata` in PHP, or somehow else set the `web_stories_tracking_optin` flag. Example using WP-CLI:

```bash
# Enable tracking for user with ID 1
$ wp user meta update 1 web_stories_tracking_optin true

# Disable tracking for user with ID 1
$ wp user meta update 1 web_stories_tracking_optin false
```

In the future, there will be a UI option for this.

**What is currently being tracked?**

Only a handful of things, for testing...

Dashboard:

* Opening the dashboard ([Page views](https://developers.google.com/analytics/devguides/collection/gtagjs/pages))
* Navigating to different screens within the dashboard ([Screen views](https://developers.google.com/analytics/devguides/collection/gtagjs/screens))  
  Examples:  
    - Navigating from "My Stories" to "Explore Templates"
    - Navigating between individual templates
* Performing actions  ([Events](https://developers.google.com/analytics/devguides/collection/gtagjs/events))
   * When clicking on "Create new story"
   * When clicking on "Use template" (sending along the template ID)

Editor:

* Opening the editor ([Page views](https://developers.google.com/analytics/devguides/collection/gtagjs/pages))
* Performing actions  ([Events](https://developers.google.com/analytics/devguides/collection/gtagjs/events))
    * When publishing a story
    * When previewing a story
    * When scheduling a story to be published in the future

## Relevant Technical Choices

Minimal logic on the PHP side. Nearly everything is done in JS, just the data is provided via PHP.

`trackEvent` & `trackScreenView` return promises and emit warnings when tracking took to long.

Inspired by Site Kit.

## To-do

* [ ] Address to-dos in code
* [ ] Define what to track and how



## User-facing changes

None. But for the sake of this PR, I am adding some screenshots to demonstrate the functionality:

<details><summary>Screenshots</summary>
<img width="1377" alt="Screenshot 2020-06-23 at 13 13 36" src="https://user-images.githubusercontent.com/841956/85403298-68058680-b55d-11ea-8205-9d5ded97e425.png">
<img width="515" alt="Screenshot 2020-06-22 at 19 00 24" src="https://user-images.githubusercontent.com/841956/85403305-6a67e080-b55d-11ea-99c4-7cb557e71f9f.png">
<img width="916" alt="Screenshot 2020-06-22 at 18 59 51" src="https://user-images.githubusercontent.com/841956/85403308-6b007700-b55d-11ea-8d12-f4a5d65df0fb.png">
</details>

## Testing Instructions

1. Opt into tracking.
1. Browse editor & dashboard
1. See tracking events in console

---



Fixes #2531
